### PR TITLE
[GEOS-10406] Make OGC API Queryables Configurable and Indexable

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-10406](https://osgeo-org.atlassian.net/browse/GEOS-10406)

…llections features

Added configurable queryables to filter and started on tests

Changed so visitor keeps track of attributes missing from queryable configuration

added globals for queryables

added queryables test data elements for the CollectionsController tests

added global queryables to oseo wicket page

Setup REST controller for global oseo setting changes

added eventing system

Test for insert create indices

Got type detection working

switched to type detection using the sample

general index

added numeric and text btree for json

added delete event handler

cleanup

switched to propertymapper

added logging

some test cleanup

more cleanup

Was grabbing the wrong schema for column names

  

# Checklist

- [ x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x ] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ x] Bug fixes and small new features are presented as a single commit.
- [x ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/ItemsTest.java [hunk 1]
```diff
@@ -145,7 +145,7 @@ public void testSentinelItemsHTML() throws Exception {
     }
 
     private void testSentinel2SampleHTML(Elements elements) {
-        assertTextContains(elements, "[data-tid='gbounds']", "-119,174, 33,333, -117,969, 34,338.");
+        assertTextContains(elements, "[data-tid='gbounds']", "-119.174, 33.333, -117.969, 34.338.");
```
Comment (aaime) [inline src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/ItemsTest.java:148]: Uh oh, another locale issue. Same as above, make the test Locale independent by setting a default locale at the begining of the test (possibly in the upper-most test class of this module).

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/ItemsTest.java [hunk 2]
```diff
@@ -491,4 +491,14 @@ public void testQueryByDynamicProperty() throws Exception {
 
         assertThat((List<String>) doc.read("features[*].id"), contains("SAS1_20180226102021.01"));
     }
+
+    @Test
+    public void testQueryByDynamicPropertyNonQueryable() throws Exception {
```
Comment (aaime) [inline src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/ItemsTest.java:496]: Good, maybe then add it and try again?

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACQueryableIndexTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
+package org.geoserver.ogcapi.stac;
```
Comment (aaime) [inline src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACQueryableIndexTest.java:1]: Copyright header

**Review by aaime — CHANGES_REQUESTED** [review]
In addition to the comments below, two minor things:
- QA failure is not your fault, rebase and it should be fixed
- The commit message is not "[GEOS-10406] Make OGC API Queryables Configurable and Indexable" (but I can fix it also while merging, with a squash-merge)

File: src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/OseoEvent.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package org.geoserver.opensearch.eo;
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/OseoEvent.java:5]: Fixed

File: src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/OseoEvent.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.List;
+import org.opengis.feature.type.Name;
+
+public class OseoEvent {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/OseoEvent.java:13]: Fixed

File: src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/OseoEventListener.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package org.geoserver.opensearch.eo;
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/OseoEventListener.java:5]: Fixed

File: src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/OseoEventType.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
+package org.geoserver.opensearch.eo;
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/OseoEventType.java:5]: Fixed

File: src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java [hunk 8]
```diff
@@ -467,6 +495,259 @@ public FeatureType getSchema(Name name) throws IOException {
...
+                                    + e.getMessage());
+                }
+                delegate.closeSafe(cx);
+            }
+        }
+        return out;
+    }
+
+    private String getIndexField(Expression expression) {
+        String out = expression.toString();
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java:644]: Fixed

File: src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java [hunk 9]
```diff
@@ -467,6 +495,259 @@ public FeatureType getSchema(Name name) throws IOException {
...
+
+    private String getIndexTitle(String fieldName) {
+        return fieldName.replaceAll(":", "_") + "_idx";
+    }
+
+    private String encodeJsonPointer(Function jsonPointer, String type) throws IOException {
+        StringBuffer out = new StringBuffer();
+        Expression json = getParameter(jsonPointer, 0, true);
+        Expression pointer = getParameter(jsonPointer, 1, true);
+        if (json instanceof PropertyName && pointer instanceof Literal) {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java:678]: Fixed

File: src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACOseoListener.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
+package org.geoserver.ogcapi.stac;
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACOseoListener.java:5]: Fixed

File: src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACOseoListener.java [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    @Override
+    public void dataStoreChange(OseoEvent event) {
+        switch (event.getType().name()) {
+            case "PostInsert":
+                handleCollectionsInsertEvent(event.getLayerName());
+                break;
+            case "PreInsert":
+                handleCollectionsInsertEvent(event.getLayerName());
+                break;
+            case "PreUpdate":
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACOseoListener.java:73]: Fixed

File: src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACOseoListener.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        }
+    }
+
+    private String getFieldType(Feature sampleFeature, Expression expression) {
+        if (!(expression instanceof JsonPointerFunction)) {
+            return NOT_JSON;
+        }
+        JsonPointerFunction jsonPointerFunction = (JsonPointerFunction) expression;
+        checkAndFixJsonPointerWithoutLeadingSlash(jsonPointerFunction);
+        String out = expression.evaluate(sampleFeature).getClass().getSimpleName();
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACOseoListener.java:155]: Fixed to default to text.

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/ItemsTest.java (hunk 1 above)
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/ItemsTest.java:152]: Fixed by setting Locale in parent.

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/ItemsTest.java (hunk 2 above)
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/ItemsTest.java:500]: Fixed

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACQueryableIndexTest.java (hunk 3 above)
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACQueryableIndexTest.java:5]: Fixed

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACQueryableIndexTest.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    public void testCreateAndDeleteIndices() throws Exception {
+        STACOseoListener stacOseoListener = getStacOseoListener();
+        List<String> indices = stacOseoListener.getIndexListByTable("product");
+        assertEquals(49, indices.size());
+        OseoEvent oseoEvent = new OseoEvent();
+        oseoEvent.setType(OseoEventType.POST_INSERT);
+        oseoEvent.setLayerName("SAS1");
+        stacOseoListener.dataStoreChange(oseoEvent);
+        indices = stacOseoListener.getIndexListByTable("product");
+        assertEquals(61, indices.size());
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACQueryableIndexTest.java:67]: Fixed by changing index naming to incorporate collection and expression

File: doc/en/api/1.0.0/owsservices.yaml [hunk 14]
```diff
@@ -1655,3 +1655,193 @@ definitions:
       verbose:
         type: boolean
         description: Flag indicating if the service should be verbose
+
+  /services/oseo/settings:
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/api/1.0.0/owsservices.yaml]: fixed

File: src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java [hunk 15]
```diff
@@ -467,6 +495,259 @@ public FeatureType getSchema(Name name) throws IOException {
...
+                        preparedStatement =
+                                cx.prepareStatement(
+                                        "CREATE INDEX IF NOT EXISTS "
+                                                + getIndexTitle(fieldName)
+                                                + " ON product ("
+                                                + getIndexField(expression)
+                                                + ")");
+                        preparedStatement.execute();
+                        break;
+                    case "String":
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java]: fixed

File: src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java [hunk 16]
```diff
@@ -467,6 +495,259 @@ public FeatureType getSchema(Name name) throws IOException {
...
+     * @param layerName Collection Type
+     * @param expression Expression that defines queryable from template
+     * @param fieldType Integer, Float, String, or NOT_JSON
+     */
+    @Override
+    public void createIndex(
+            String fieldName, String layerName, Expression expression, String fieldType) {
+        JDBCDataStore delegate = getRawDelegateStore();
+        SQLDialect dialect = delegate.getSQLDialect();
+        if (dialect instanceof PostGISDialect) {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java]: fixed

File: src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java [hunk 17]
```diff
@@ -467,6 +495,259 @@ public FeatureType getSchema(Name name) throws IOException {
...
+            Connection cx = null;
+            PreparedStatement preparedStatement = null;
+            try {
+                cx = delegate.getConnection(Transaction.AUTO_COMMIT);
+                switch (fieldType) {
+                    case NOT_JSON:
+                        preparedStatement =
+                                cx.prepareStatement(
+                                        "CREATE INDEX IF NOT EXISTS "
+                                                + getIndexTitle(fieldName)
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/JDBCOpenSearchAccess.java]: fixed

**Comment by aaime** [comment]
QA failing, not formatted correctly. Please use "mvn install" before pushing a PR, it fixes the formatting of java files and pom.xml files.

**Pull request merged**